L42-python-metadata-flags.md

```diff
@@ -0,0 +1,153 @@
+Surface 'wait-for-ready' setting in Python client
+----
+* Author(s): lidizheng
+* Approver: srini100
+* Status: In Review
+* Implemented in: Python
+* Last updated: October 16, 2018
+* Discussion at: https://groups.google.com/forum/#!topic/grpc-io/Q6fgbn0RZLs
+
+## Abstract
+
+Surface 'wait-for-ready' configuration in gRPC Python client that will use the mechanism provided by C-Core via the initial metadata flags. The implementation should allow adding future such configurations in gRPC Python easily.
+
+## Background
+
+### Definition of 'wait-for-ready' semantics
+> If an RPC is issued but the channel is in TRANSIENT_FAILURE or SHUTDOWN states, the RPC is unable to be transmitted promptly. By default, gRPC implementations SHOULD fail such RPCs immediately. This is known as "fail fast," but the usage of the term is historical. RPCs SHOULD NOT fail as a result of the channel being in other states (CONNECTING, READY, or IDLE).
+> 
+> gRPC implementations MAY provide a per-RPC option to not fail RPCs as a result of the channel being in TRANSIENT_FAILURE state. Instead, the implementation queues the RPCs until the channel is READY. This is known as "wait for ready." The RPCs SHOULD still fail before READY if there are unrelated reasons, such as the channel is SHUTDOWN or the RPC's deadline is reached.
+> 
+> From https://github.com/grpc/grpc/blob/master/doc/wait-for-ready.md 
+
+### Use cases for 'wait-for-ready'
+
+When developers spin up gRPC clients and servers in the same time, it is very like to fail first couple RPC calls due to unavailability of the server. If developers failed to prepare for this situation, the result can be catastrophic. But with 'wait-for-ready' semantics, developers can initialize the client and server in any order, especially useful in testing.
+
+Also, developers may ensure the server is up before starting client. But in some cases like transient network failure may result in a temporary unavailability of the server. With 'wait-for-ready' semantics, those RPC calls will automatically wait until the server is ready to accept incoming requests.
+
+### Current 'wait-for-ready' behavior
+
+By default, the client RPC call will fail immediately if the server is unavailable.
+
+But there is a Python-only function to simulate the 'wait-for-ready' mechanism. It is a class named `_ChannelReadyFuture` in `_utilities.py`. It will create a `Future` object to enable the client to block till the channel is ready. It works just like the 'wait-for-ready' mechanism with Pythonic handling approach, but it has three drawbacks:
```

Comment:
Point 3 below (it only applies to the first connection attempt) seemingly contradicts that this works just like C-core's wait-for-ready mechanism.

Reply:
Yes. The ChannelReadyFuture mechanism gRPC Python have now is not using the C-Core's wait-for-ready mechanism, but rely on `watch_connectivity_state` and `check_connectivity_state` APIs. That is one of the incentives to implement this feature.

Reply:
I would change this to "It works similarly to...". Saying "just like" indicates that they have exactly the same behavior.